### ISSUE TITLE
Bad rendering on quick rotation

### ISSUE THREAD

**Issue by eugeneloza** [eugeneloza/decoherence#161]
Even with `Window.DobuleBuffer = true` I get wierd flicker of not completely rendered scene.
I have to make a manual render container (in background) and average image outputs? That'd be nice depending on speed. Some games use 4 frames averaging.
See https://sourceforge.net/p/castle-engine/discussion/general/thread/8748aa06/ and http://www.gamedev.ru/flame/forum/?id=205199

**Comment by michaliskambi** [comment]
I also observed the flickering in Decoherence on Linux, and forgot to report it earlier... Sorry! Some hints:

I do not observe it in any other CGE games, and it definitely should *not* happen if you don't do any "dirty tricks" with the engine :)

Some ideas what to check in Decoherence code:
- Are you using any direct OpenGL commands?
- Are you using threads and accidentally making some rendering from a thread? If you have the option to "run Decoherence in a mode without any threads", then this would be easy to check.
- Do you render your UI controls consistently in all the frames? You don't toggle the Exists property of some 2D controls every frame, or otherwise make the UI just invisible every other frame?

You should *not* need to add any complicated rendering trick to workaround this. In my opinion, rendering to a special background container and averaging 2 or 4 frames would be a bad direction here, as 1. it will noticeably lower your game FPS, and 2. it's also non-trivial to implement (in an efficient manner) unless you already used TGLRenderToTexture. Such tricks make sense if you want a fancy screen effect (motion blur), but they should not be used when the underlying issue is most probably an ordinary bug in either Decoherence or Castle Game Engine :)

Bottom line is: There's bug somewhere, it should be found :) Either you do something "dirty" in Decoherence code, and cause undefined behavior from the OpenGL rendering, or there's a bug in CGE. It should be found and fixed :)

**Comment by eugeneloza** [comment]
> Are you using any direct OpenGL commands?

No. Every render operation is done through Castle Game Engine in `Window.onRender`. And if I disable absolutely everything (just commenting out this line https://github.com/eugeneloza/decoherence/blob/master/src/decoherence.pas#L132 will disable all GUI rendering - it's the only thing I'm "drawing" manually). The flicker is still there.

> Are you using threads and accidentally making some rendering from a thread? 

No, I'm not using any threads during the game at the moment (not even timer). And loading in a thread is disabled (only LoadScreen images are loaded in a thread). Yet, I didn't make the compiler directive to yet, it's still in todo #126 

> Do you render your UI controls consistently in all the frames? 

Yes, GUI renders at every `Window.onRender`, called at the same line https://github.com/eugeneloza/decoherence/blob/master/src/decoherence.pas#L132 And the flickering doesn't go away if completely disable its rendering by commenting this line out

> if you want a fancy screen effect

Yes, that's what I'm working on :) You can see I've separated a CameraMan actor to provide for rotation innertia, shaking the screen, friction and other similar softening effects. Practically it's a reimplementation of some TWalkCamera routines. I also need a flexible camera for combat where user input should be handled together with code-controlled camera (turning at each enemy on its attack).

> lower your game FPS

Yes, of course. I'm still thinking on properly/efficiently implementing this.
Thou the flickering is caused because I simultaneously see images from 3 to 5 frames (if I shake the screen near some distinctive element, like a contrast wall edge) at ~80 FPS, and all images are contrast enough to be seen.

> TGLRenderToTexture

Yes, I've been using these before, it worked nice for me :) That was exactly the way I was thinking of.

**Comment by michaliskambi** [comment]
> (not even timer)

Using timer (Application.OnTimer, TCastleTimer, or Lazarus timer component) is not a problem (with  respect to threads). Timer is not executed in a separate thread, it's executed in the main thread by the CastleWindow or Lazarus application event loop. So it never runs in parallel to other code in the main thread.  So you're free to call any LCL or Castle Game Engine functions inside the timer, guaranteed :)

> Thou the flickering is caused because I simultaneously see images from 3 to 5 frames (if I shake the screen near some distinctive element, like a contrast wall edge) at ~80 FPS, and all images are contrast enough to be seen.

I'm not sure what do you mean by shaking the screen, do you shake the GUI in Decoherence? Do you mean that you know why the flickering happens?

Otherwise, I would really like to debug and fix it, regardless whether you make a cool graphic effect that by the way also "hides" the problem. Making a cool graphic effect is of course OK, but it should not be used to "workaround" a bug, if it's something incorrect in Decoherence or CGE :) If this is the case, I'll see in a few days can I help to investigate it -- a rendering bug may as well be inside CGE, so I'm very interested in catching it:)

**Comment by eugeneloza** [comment]
> Using timer (Application.OnTimer, TCastleTimer, or Lazarus timer component) is not a problem

I already implemented a threaded timer based on these (as in https://github.com/eugeneloza/decoherence/issues/125 ) here https://github.com/eugeneloza/decoherence/blob/master/src/decotime.pas . But it's never called yet (designed to be called from World.Manage in Window.onBeforeRender).

> I'm not sure what do you mean by shaking the scree

Shaking the mouse wildly with `MouseLook = true` :) The same issue happens in a normal game when quickly rotating over turns. Looks like I can see the same (but less aggressive due to softer textures) flickering in `fps_game` example with 4 to 5 images rendered in one frame. P.S. as far as I remember the effect is softer on Linux.

**Comment by michaliskambi** [comment]
In fps_game, I don't see any flickering, ever. No matter how hardly I shake my mouse :)

In Decoherence, I can see the user interface flickering without moving the mouse at all. Even at loading screen, it's flickering, there's a dark strip moving through the screen. It definitely seems like the DoubleBuffer is turned off.

...and, doing a grep, it was trivial to find a line:

> window.DoubleBuffer := false;//true;             //what's the difference? speed? memory?

inside decoherence.pas (line 309)... You turn off the double-buffering explicitly, and that's the cause of the flickering. You're drawing straight into the (visible) screen, so user sees as you're overdrawing the UI every frame. Simply remove this line (DoubleBuffer is true by default) and the problem is solved :)

This flickering is exactly the problem that double-buffering is solving. For all practical purposes, you always want the double-buffering turned "on", it does not cost you speed or memory, and that's why it's turned on by default. (Well, it does, internally, use an extra buffer at the GPU side. But this buffer is most probably specifically invented for the double-buffering purposes as far as I know, so you're not "losing" anything. 100% of games (and non-games) you play are using double-buffering :)

**Comment by eugeneloza** [comment]
> You turn off the double-buffering

Emmm... it should be "on" in latest versions... https://github.com/eugeneloza/decoherence/blob/master/src/decoherence.pas#L321 I don't remember when I turned it on, maybe not long before... :)
And actually I never saw "the user interface flickering"...

Hmm... maybe that's another trick of my dying Radeon 5400 card?

I wonder, if I had `DoubleBuffer` enabled for this video? https://www.youtube.com/watch?v=ZKeMQah3sEA (it has flickering recorded).
I've recorded a video (with `DoubleBuffer = true`) and watched it frame-by-frame. Doesn't seem like anything is going wrong there - all frames are perfectly rendered, thou I still see the flickering _in situ_. Hmm... then it might be reasonable to blame my cheap and old video card. I don't remember any flickering at my work PC (thou I did not test it thoroughly, only compiled to see if its working).
Doesn't look like anything can be done here, therefore. Except buying a game-class videocard sometime...
Still I wonder why other games don't show any significant flicker? Maybe, it's difference between OpenGL and DirectX (or rather their interaction with card driver)? I remember I've had some effects missing on a 2D game, which (afair) used OpenGL.

Well... anyway, there is hardly anything objective left here at the moment. So, let's blame my videocard and close the issue :D At least for now.
Thanks a lot for your help, @michaliskambi ! I should have been more attentive :)

**Comment by michaliskambi** [comment]
> Emmm... it should be "on" in latest versions... 
> https://github.com/eugeneloza/decoherence/blob/master/src/decoherence.pas#L321 I don't remember when I turned it on, maybe not long before... :)
> And actually I never saw "the user interface flickering"...

Ups, you're right, I didn't update Decoherence sources. With the latest version, that has DoubleBuffer = true, I don't see any flickering on my Linux.

I'm very sorry -- all this time I was talking about a problem that was fixed long time ago...

I marked the Decoherence repository now on my system to have it automatically updated (I use "myrepos" tool to manage my GIT and SVN repositories.)

I don't see flickering on https://www.youtube.com/watch?v=ZKeMQah3sEA either :)

**Comment by eugeneloza** [comment]
> I don't see flickering

Well... then I really should consider buying a normal videocard some day... damn, they're expensive :D Thou, I believe my old one will soon give me a nice push, when it'll finally stop working at all...

**Comment by michaliskambi** [comment]
> I don't see flickering on https://www.youtube.com/watch?v=ZKeMQah3sEA either :)

Watching it more closely, I don't see a flickering, but I think the "vertical synchronization" is off on your GPU. So the UI is never flickering, but sometimes there's like a "tear" visible when the screen is redrawn, and the 3D contents change significantly.

Well, this is indeed a problem on the GPU. We could possibly force (or at least suggest) to use vertical sync from the engine, but from what I know most modern games don't do it -- since modern GPUs have it automatically "on".

So, I think I can see your artifact in the movie -- and, if we're speaking about the same thing, then 1. it is indeed related to your GPU, 2. but it can be configurable to better. Look for some option related "vertical sync" in some advanced settings of GPU. It is possible that Windows and Linux drivers have different defaults for it, too, so it's possible that you only see the problem on one system.
